synapse/storage/end_to_end_keys.py

```diff
@@ -281,3 +285,145 @@ def delete_e2e_keys_by_device_txn(txn):
         return self.runInteraction(
             "delete_e2e_keys_by_device", delete_e2e_keys_by_device_txn
         )
+
+    def _set_e2e_device_signing_key_txn(self, txn, user_id, key_type, key):
+        """Set a user's cross-signing key.
+
+        Args:
+            txn (twisted.enterprise.adbapi.Connection): db connection
+            user_id (str): the user to set the signing key for
+            key_type (str): the type of key that is being set: either 'master'
+                for a master key, 'self_signing' for a self-signing key, or
+                'user_signing' for a user-signing key
+            key (dict): the key data
+        """
+        # the cross-signing keys need to occupy the same namespace as devices,
+        # since signatures are identified by device ID.  So add an entry to the
+        # device table to make sure that we don't have a collision with device
+        # IDs
+        for v in key["keys"].values():
+            pubkey = v
+            break
+        self._simple_insert(
+            "devices",
+            values={
+                "user_id": user_id,
+                "device_id": pubkey,
+                "display_name": key_type + " signing key",
+                "hidden": True
+            },
+            desc="store_master_key_device"
+        )
+
+        # and finally, store the key itself
+        self._simple_insert(
+            "e2e_device_signing_keys",
+            values={
+                "user_id": user_id,
+                "keytype": key_type,
+                "keydata": json.dumps(key),
+                "ts": time.time() * 1000
+            },
+            desc="store_master_key"
+        )
+
+    def set_e2e_cross_signing_key(self, user_id, key_type, key):
```

Comment:
there seems to be some confusion here about whether it is a *device* signing key or a *cross* signing key.